Wrappers/Python/cil/optimisation/functions/IndicatorBox.py

```diff
@@ -29,67 +35,389 @@ class IndicatorBox(Function):
+        if backend == 'numba':
 
-    def __call__(self,x):
-        
-        '''Evaluates IndicatorBox at x'''
-                
-        if (numpy.all(x.as_array() >= self.lower) and 
-            numpy.all(x.as_array() <= self.upper) ):
-            val = 0
+            logging.info("Numba backend is used.")      
+            return super(IndicatorBox, cls).__new__(IndicatorBox_numba)
         else:
-            val = numpy.inf
-        return val
+            logging.info("Numpy backend is used.") 
+            return super(IndicatorBox, cls).__new__(IndicatorBox_numpy)
     
-    def gradient(self,x):
-        return ValueError('Not Differentiable') 
-    
-    def convex_conjugate(self,x):
+    def __init__(self, lower=None, upper=None, backend='numba'):
+        '''__init__'''
+        super(IndicatorBox, self).__init__()
         
-        '''Convex conjugate of IndicatorBox at x'''
+        # We set lower and upper to either a float or a numpy array        
+        self.lower = -np.inf if lower is None else _get_as_nparray_or_number(lower)
+        self.upper =  np.inf if upper is None else _get_as_nparray_or_number(upper)
 
-        return x.maximum(0).sum()
-         
-    def proximal(self, x, tau, out=None):
+        self.orig_lower = lower
+        self.orig_upper = upper
+        # default is to evaluate the function
+        self._suppress_evaluation = False
+
+    @property
+    def suppress_evaluation(self):
+        return self._suppress_evaluation
+
+    def set_suppress_evaluation(self, value):
+        '''Suppresses the evaluation of the function
         
-        r'''Proximal operator of IndicatorBox at x
+        Parameters
+        ----------
 
-            .. math:: prox_{\tau * f}(x)
+            value : bool
+                If True, the function evaluation on any input will return 0, without calculation.
         '''
+        if not isinstance(value, bool):
+            raise ValueError('Value must be boolean')
+        self._suppress_evaluation = value
+
+    def __call__(self,x):
+        '''Evaluates IndicatorBox at x
         
-        if out is None:
-            return (x.maximum(self.lower)).minimum(self.upper)        
-        else:               
-            x.maximum(self.lower, out=out)
-            out.minimum(self.upper, out=out) 
+        Parameters
+        ----------
+        
+            x : DataContainer
             
+        Evaluates the IndicatorBox at x. If ``suppress_evaluation`` is ``True``, returns 0.  
+        '''
+        if not self.suppress_evaluation:
+            return self.evaluate(x)    
+        return 0.0
+    
     def proximal_conjugate(self, x, tau, out=None):
         
         r'''Proximal operator of the convex conjugate of IndicatorBox at x:
 
-          ..math:: prox_{\tau * f^{*}}(x)
+          .. math:: prox_{\tau * f^{*}}(x)
+
+          Parameters
+          ----------
+
+            x : DataContainer
```

Comment:
indent levels

Reply:
I think this is required to render correctly

Reply:
It's not the same as elsewhere.

https://github.com/TomographicImaging/CIL/blob/ffe7449a0dd675687b7f357bea89e45ef8f42d7b/Wrappers/Python/cil/optimisation/algorithms/PDHG.py#L31-L36